In [1]:
from google.colab import drive
drive.mount("/content/gdrive/")

%cd "/content/gdrive/My Drive/Colab Notebooks/bikes-theft-model"

# Libraries: Standard ones
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.io as pio

# Library for boxplots
import seaborn as sns
import pandas as pd

#GRAPHS CLASS
from Codigos.DataStatistics import GraphsStatistics as gp


Mounted at /content/gdrive/
/content/gdrive/My Drive/Colab Notebooks/bikes-theft-model


/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning:

pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.



# **Data sets description**

## **Thefts data sets**

INTRO

### **Toronto data set**

This dataset contains Bicycle Thefts occurrences from **2014-2019** . The location of crime occurrences have been deliberately offset to the nearest road intersection node to protect the privacy of parties involved in the occurrence. All location data must be considered as an approximate location of the occurrence and users are advised not to interpret any of these locations as related to a specific address or individual, it has a total of 26 features. The data set can be found [here](https://https://data.torontopolice.on.ca/datasets/698717de582845458159a73ad4a2a8e0_0?geometry=-80.680%2C43.538%2C-78.066%2C43.885)

<table style="undefined;table-layout: fixed; width: 726px">
<colgroup>
<col style="width: 154px">
<col style="width: 375px">
<col style="width: 110px">
<col style="width: 87px">
</colgroup>
<thead>
  <tr>
    <th>Field</th>
    <th>Field Description</th>
    <th>Variable Type</th>
    <th>Num of Unique values</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>X</td>
    <td>Location in cartetian coordinates (X)</td>
    <td>float</td>
    <td>4885</td>
  </tr>
  <tr>
    <td>Y</td>
    <td>Location in cartetian coordinates (Y)</td>
    <td>float</td>
    <td>4874</td>
  </tr>
  <tr>
    <td>FID</td>
    <td>ID </td>
    <td>int</td>
    <td>21584</td>
  </tr>
  <tr>
    <td>Index</td>
    <td>Record Unique Identifier</td>
    <td>int</td>
    <td>21584</td>
  </tr>
  <tr>
    <td>event_unique_id</td>
    <td>Event Occurrence Identifier</td>
    <td>String</td>
    <td>19350</td>
  </tr>
  <tr>
    <td>Primary_Offence</td>
    <td>Offence related to the occurrence</td>
    <td>String</td>
    <td>65</td>
  </tr>
  <tr>
    <td>Occurrence_Date</td>
    <td>Date of occurrence</td>
    <td>String</td>
    <td>2104</td>
  </tr>
  <tr>
    <td>Occurrence_Year</td>
    <td>Occurrence year</td>
    <td>int</td>
    <td>6</td>
  </tr>
  <tr>
    <td>Occurrence_Month</td>
    <td>Occurrence Month</td>
    <td>int</td>
    <td>12</td>
  </tr>
  <tr>
    <td>Occurrence_Day</td>
    <td>Occurrence Day</td>
    <td>int</td>
    <td>31</td>
  </tr>
  <tr>
    <td>Occurrence_Time</td>
    <td>Occurrence Time</td>
    <td>String</td>
    <td>933</td>
  </tr>
  <tr>
    <td>Division</td>
    <td>Police Division where event occurred</td>
    <td>int</td>
    <td>18</td>
  </tr>
  <tr>
    <td>City</td>
    <td>City where event occurred</td>
    <td>String</td>
    <td>1</td>
  </tr>
  <tr>
    <td>Location_Type</td>
    <td>Location Type where event occurred</td>
    <td>String</td>
    <td>44</td>
  </tr>
  <tr>
    <td>Premise_Type</td>
    <td>Premise Type where event occurred</td>
    <td>String</td>
    <td>5</td>
  </tr>
  <tr>
    <td>Bike_Make</td>
    <td>Bicycle Make</td>
    <td>String</td>
    <td>725</td>
  </tr>
  <tr>
    <td>Bike_Model</td>
    <td>Bicycle Model</td>
    <td>String</td>
    <td>7008</td>
  </tr>
  <tr>
    <td>Bike_Type</td>
    <td>Bicycle Type</td>
    <td>String</td>
    <td>13</td>
  </tr>
  <tr>
    <td>Bike_Speed</td>
    <td>Bicycle Speed</td>
    <td>int</td>
    <td>62</td>
  </tr>
  <tr>
    <td>Bike_Colour</td>
    <td>Bicycle Colour</td>
    <td>String</td>
    <td>233</td>
  </tr>
  <tr>
    <td>Cost_of_Bike</td>
    <td>Cost of Bicycle</td>
    <td>float</td>
    <td>1458</td>
  </tr>
  <tr>
    <td>Status</td>
    <td>Status of event</td>
    <td>String</td>
    <td>3</td>
  </tr>
  <tr>
    <td>Hood_ID</td>
    <td>Neighbourhood Id</td>
    <td>int</td>
    <td>140</td>
  </tr>
  <tr>
    <td>Neighbourhood</td>
    <td>Neighbourhood name</td>
    <td>String</td>
    <td>140</td>
  </tr>
  <tr>
    <td>Lat</td>
    <td>Longitude of point extracted after offsetting X and &amp; Coordinates to&nbsp;&nbsp;&nbsp;nearest intersection node</td>
    <td>float</td>
    <td>4874</td>
  </tr>
  <tr>
    <td>Long</td>
    <td>Latitude of point extracted after offsetting X and &amp; Coordinates to&nbsp;&nbsp;&nbsp;nearest intersection node</td>
    <td>float</td>
    <td>4885</td>
  </tr>
</tbody>
</table>

### **Ottawa data set**


This dataset contains Bicycle Thefts occurrences from **2015-2019** . The location of crime occurrences have been deliberately offset to the nearest road intersection node to protect the privacy of parties involved in the occurrence. All location data must be considered as an approximate location of the occurrence and users are advised not to interpret any of these locations as related to a specific address or individual, it has a total of 19 features. The data set can be found [here](https://https://open.ottawa.ca/datasets/3620cc7a3b874557bb288d889a4d56c2_0?geometry=-78.421%2C45.008%2C-73.194%2C45.683)

<table style="undefined;table-layout: fixed; width: 726px">
<colgroup>
<col style="width: 154px">
<col style="width: 375px">
<col style="width: 110px">
<col style="width: 87px">
</colgroup>
<thead>
  <tr>
    <th>Field</th>
    <th>Field Description</th>
    <th>Variable Type</th>
    <th>Num of Unique values</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>X</td>
    <td>Location in cartetian coordinates (X)</td>
    <td>float</td>
    <td>4885</td>
  </tr>
  <tr>
    <td>Y</td>
    <td>Location in cartetian coordinates (Y)</td>
    <td>float</td>
    <td>4874</td>
  </tr>
  <tr>
    <td>FID</td>
    <td>ID </td>
    <td>int</td>
    <td>21584</td>
  </tr>
  <tr>
    <td>Index</td>
    <td>Record Unique Identifier</td>
    <td>int</td>
    <td>21584</td>
  </tr>
  <tr>
    <td>event_unique_id</td>
    <td>Event Occurrence Identifier</td>
    <td>String</td>
    <td>19350</td>
  </tr>
  <tr>
    <td>Primary_Offence</td>
    <td>Offence related to the occurrence</td>
    <td>String</td>
    <td>65</td>
  </tr>
  <tr>
    <td>Occurrence_Date</td>
    <td>Date of occurrence</td>
    <td>String</td>
    <td>2104</td>
  </tr>
  <tr>
    <td>Occurrence_Year</td>
    <td>Occurrence year</td>
    <td>int</td>
    <td>6</td>
  </tr>
  <tr>
    <td>Occurrence_Month</td>
    <td>Occurrence Month</td>
    <td>int</td>
    <td>12</td>
  </tr>
  <tr>
    <td>Occurrence_Day</td>
    <td>Occurrence Day</td>
    <td>int</td>
    <td>31</td>
  </tr>
  <tr>
    <td>Occurrence_Time</td>
    <td>Occurrence Time</td>
    <td>String</td>
    <td>933</td>
  </tr>
  <tr>
    <td>Division</td>
    <td>Police Division where event occurred</td>
    <td>int</td>
    <td>18</td>
  </tr>
  <tr>
    <td>City</td>
    <td>City where event occurred</td>
    <td>String</td>
    <td>1</td>
  </tr>
  <tr>
    <td>Location_Type</td>
    <td>Location Type where event occurred</td>
    <td>String</td>
    <td>44</td>
  </tr>
  <tr>
    <td>Premise_Type</td>
    <td>Premise Type where event occurred</td>
    <td>String</td>
    <td>5</td>
  </tr>
  <tr>
    <td>Bike_Make</td>
    <td>Bicycle Make</td>
    <td>String</td>
    <td>725</td>
  </tr>
  <tr>
    <td>Bike_Model</td>
    <td>Bicycle Model</td>
    <td>String</td>
    <td>7008</td>
  </tr>
  <tr>
    <td>Bike_Type</td>
    <td>Bicycle Type</td>
    <td>String</td>
    <td>13</td>
  </tr>
  <tr>
    <td>Bike_Speed</td>
    <td>Bicycle Speed</td>
    <td>int</td>
    <td>62</td>
  </tr>
  <tr>
    <td>Bike_Colour</td>
    <td>Bicycle Colour</td>
    <td>String</td>
    <td>233</td>
  </tr>
  <tr>
    <td>Cost_of_Bike</td>
    <td>Cost of Bicycle</td>
    <td>float</td>
    <td>1458</td>
  </tr>
  <tr>
    <td>Status</td>
    <td>Status of event</td>
    <td>String</td>
    <td>3</td>
  </tr>
  <tr>
    <td>Hood_ID</td>
    <td>Neighbourhood Id</td>
    <td>int</td>
    <td>140</td>
  </tr>
  <tr>
    <td>Neighbourhood</td>
    <td>Neighbourhood name</td>
    <td>String</td>
    <td>140</td>
  </tr>
  <tr>
    <td>Lat</td>
    <td>Longitude of point extracted after offsetting X and &amp; Coordinates to&nbsp;&nbsp;&nbsp;nearest intersection node</td>
    <td>float</td>
    <td>4874</td>
  </tr>
  <tr>
    <td>Long</td>
    <td>Latitude of point extracted after offsetting X and &amp; Coordinates to&nbsp;&nbsp;&nbsp;nearest intersection node</td>
    <td>float</td>
    <td>4885</td>
  </tr>
</tbody>
</table>

## **Census data set**

These 2016 Census data set provide statistical information about the population, age and sex, type of dwelling, families, households and marital status, language among other features. In general terms, the census information is segmented into different geographical units, such as: Census Metropolitan Areas (CMA), Census Division (CD), Census Branch (CSD), Census Area (CT) and Dissemination Area (DA).  For this study, the CT, which is a small and relatively stable area, will be taken as the geographic unit. Census areas usually have a population of 2,500 to 8,000 people. They are located in large urban centers that should have a central urban population of 50,000 or more. Data was obtained from  [[1]](https://https://www12.statcan.gc.ca/census-recensement/2016/dp-pd/index-eng.cfm) and  [[2]](https://https://censusmapper.ca/) 

The following is a statistical analysis for the **cities of Toronto and Ottawa**


In the next table are shown the set of caracteristics for the population data set:

<table>
<thead>
  <tr>
    <th>Field</th>
    <th>Field description</th>
    <th>Variable Type</th>
    <th>Num of unique values</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>GeoUID</td>
    <td>Region id</td>
    <td>float64</td>
    <td>204</td>
  </tr>
  <tr>
    <td>Type</td>
    <td>Subdivision type (Census Track)</td>
    <td>String</td>
    <td>1</td>
  </tr>
  <tr>
    <td>Region Name</td>
    <td>Subdivision Name</td>
    <td>String</td>
    <td>4</td>
  </tr>
  <tr>
    <td>Area (sq km)</td>
    <td>Area in squer Kilimeters</td>
    <td>float64</td>
    <td>203</td>
  </tr>
  <tr>
    <td>Dwellings</td>
    <td>Number of houses</td>
    <td>int64</td>
    <td>196</td>
  </tr>
  <tr>
    <td>Households</td>
    <td>Number of people who live in the <br>house</td>
    <td>int64</td>
    <td>187</td>
  </tr>
  <tr>
    <td>v_CA16_401: Population, 2016</td>
    <td>Number of people in the region </td>
    <td>int64</td>
    <td>201</td>
  </tr>
  <tr>
    <td>v_CA16_406: Population density per square kilometre</td>
    <td>Population density per square <br>kilometre</td>
    <td>float64</td>
    <td>203</td>
  </tr>
  <tr>
    <td>v_CA16_407: Land area in square kilometres</td>
    <td>Land area in square kilometres</td>
    <td>float64</td>
    <td>167</td>
  </tr>
</tbody>
</table>

In the next table are shown the set of caracteristics for the number of reported bikes data set:

<table>
<thead>
  <tr>
    <th>Field</th>
    <th>Field description</th>
    <th>Variable Type</th>
    <th>Num of unique values</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>GeoUID</td>
    <td>ID of subdivision</td>
    <td>float64</td>
    <td>204</td>
  </tr>
  <tr>
    <td>Type</td>
    <td>Subdivision type <br>(Census Track)</td>
    <td>object</td>
    <td>1</td>
  </tr>
  <tr>
    <td>Region Name</td>
    <td>Region Name</td>
    <td>object</td>
    <td>4</td>
  </tr>
  <tr>
    <td>Area (sq km)</td>
    <td>Region area</td>
    <td>float64</td>
    <td>203</td>
  </tr>
  <tr>
    <td>Population</td>
    <td>Population number</td>
    <td>int64</td>
    <td>201</td>
  </tr>
  <tr>
    <td>Dwellings</td>
    <td>House number</td>
    <td>int64</td>
    <td>196</td>
  </tr>
  <tr>
    <td>Households</td>
    <td>Person number</td>
    <td>int64</td>
    <td>187</td>
  </tr>
  <tr>
    <td>v_CA16_5809: Bicycle</td>
    <td>Number of bycicle reported<br>by women</td>
    <td>object</td>
    <td>24</td>
  </tr>
  <tr>
    <td>v_CA16_5808: Bicycle</td>
    <td>Number of bycicle reported<br>by men</td>
    <td>object</td>
    <td>32</td>
  </tr>
  <tr>
    <td>v_CA16_5807: Bicycle</td>
    <td>Total number of bycicle</td>
    <td>object</td>
    <td>45</td>
  </tr>
</tbody>
</table>

#**Data joint construction**

In order to make a comparison between the thefts carried out between the cities of *Ottawa* and *Toronto*, a new data set is created, which aims to summarize the information of population, number of bicycles reported (by area $Km^{2}$) and number of stolen bicycles data given certain location.



To carry out the analysis by region in each city, the census track (CT) is taken as a subdivision. First the population and bicycle data sets are imported for each city.

In [2]:
######Importing data sets ##########
dp_ottawa=pd.read_csv('Data/Population/ontario_population_ottawa.csv',header=0) #population Ottawa per CT
dbikes_ottawa= pd.read_csv('Data/Bikes/ontario_bikes_Ottawa.csv',header=0) #Number of bikes Ottawa per CT
dp_toronto=pd.read_csv('Data/Population/toro_population_toronto.csv',header=0) #population Toronto per CT
dbikes_toronto= pd.read_csv('Data/Bikes/toto_bikes_toronto.csv',header=0) #Number of bikes Toronto per CT

#####To viasualize the general information of data sets #####
#display(dp_ottawa.info())
#dp_ottawa.nunique()
#display(dbikes_ottawa.info())
#dbikes_ottawa.nunique()

In [3]:
#####CREATE ONE MIX DATAFRAME FOR ALL DATA PER CITY ###########

Data_Ottawa_pop = dp_ottawa.merge(dbikes_ottawa, how = 'inner', left_on='GeoUID', right_on='GeoUID') #Joint data of population and bikes for ottawa
Data_toronto_pop = dp_toronto.merge(dbikes_toronto, how = 'inner', left_on='GeoUID', right_on='GeoUID') #Joint data of population and bikes for toronto


####Rename of some columns #######
Data_Ottawa_pop=Data_Ottawa_pop.rename(columns={'v_CA16_401: Population, 2016':'PopulationCT','v_CA16_406: Population density per square kilometre':'PopulationDensity',
                                                'v_CA16_407: Land area in square kilometres':'LandArea(skm)','v_CA16_5809: Bicycle':'BicycleF','v_CA16_5808: Bicycle':'BicycleM','v_CA16_5807: Bicycle':'BicycleT'})

Data_toronto_pop=Data_toronto_pop.rename(columns={'v_CA16_401: Population, 2016':'PopulationCT','v_CA16_406: Population density per square kilometre':'PopulationDensity',      
                                               'v_CA16_407: Land area in square kilometres':'LandArea(skm)','v_CA16_5809: Bicycle':'BicycleF','v_CA16_5808: Bicycle':'BicycleM','v_CA16_5807: Bicycle':'BicycleT'})

Data_Ottawa_pop.loc[132,'BicycleT']=0

Data_toronto_pop['BicycleT'] = pd.to_numeric(Data_toronto_pop['BicycleT'])

#Data_Ottawa_pop.loc[Data_Ottawa_pop['BicycleT']=='x']
Data_Ottawa_pop['BicycleT'] = pd.to_numeric(Data_Ottawa_pop['BicycleT'])
#print('Done')


Considering that the bicycle theft databases provide the approximate location of the incident, this parameter is used to unify the databases. The thefts are grouped by TC.

In [4]:
data_ottawa=pd.read_csv('Data/Bicycle_Theft_Ottawa_geo.csv',header=0) #Theft data by CT ottawa
data_toronto= pd.read_csv('Data/Bicycle_Thefts_Toronto_geo.csv',header=0) #Theft data by CT Toronto

#display(data_ottawa.info())  #General information of joint data set Ottawa
#display(data_toronto.info()) #General information of joint data set Toronto

In [5]:
#Copy for pick the must important columns
Toronto_data=data_toronto[['Index_', 'event_unique_id', 'Primary_Offence',
       'Occurrence_Date', 'Occurrence_Year', 'Occurrence_Month',
       'Occurrence_Day', 'Occurrence_Time', 'Division', 'City',
       'Location_Type', 'Premise_Type', 'Bike_Make', 'Bike_Model', 'Bike_Type',
       'Bike_Speed', 'Bike_Colour', 'Cost_of_Bike', 'Status', 'Hood_ID',
       'Neighbourhood', 'Lat', 'Long', 'id']].copy()

Ottawa_data=data_ottawa[['ID', 'Division', 'Sector', 'Report_Date', 'Incident_Date',
       'Incident_Weekday', 'Location_Type', 'Primary_Offence', 'Neighbourhood',
       'Bicycle_Status', 'Bicycle_Value', 'Bicycle_Make', 'Bicycle_Model',
       'Bicycle_Type', 'Bicycle_Frame', 'Bicycle_Colour', 'Bicycle_Speed', 'x',
       'y','id']].copy()

Toronto_data=Toronto_data.rename(columns={'id':'GeoUID'}) #Final dataframe with theft data Toronto
Ottawa_data=Ottawa_data.rename(columns={'id':'GeoUID'}) #Final dataframe with theft data Ottawa

#####Configuration of OcurrenceDate##########
day_type = pd.api.types.CategoricalDtype(categories=["Monday","Tuesday","Wednesday","Thursday","Friday","Saturday","Sunday"], ordered=True)


Toronto_data['Occurrence_Date']=pd.to_datetime(Toronto_data['Occurrence_Date']) 
Toronto_data['Week_day']=Toronto_data['Occurrence_Date'].dt.day_name() 
Toronto_data["Week_day"] = Toronto_data["Week_day"].astype(day_type)

Ottawa_data['Incident_Date']=pd.to_datetime(Ottawa_data['Incident_Date']) 
Ottawa_data['Week_day']=Ottawa_data['Incident_Date'].dt.day_name() 
Ottawa_data["Week_day"] = Ottawa_data["Week_day"].astype(day_type)


Finally, there is a single data set by city with all the summary information of population, number of bicycles and bicycles stolen by CT (Land area)

In [6]:

####### CONCATENATION OF SETS ###############
JointOttawa = pd.merge(Data_Ottawa_pop, Ottawa_data, on='GeoUID', how='right') #JOINT DATA SET OF POPULATION, # BIKES , and THEFTS PER CT
JointToronto = pd.merge(Data_toronto_pop, Toronto_data, on='GeoUID', how='right') # JOINT DATA SET OF POPULATION, # BIKES , and THEFTS PER CT

JointToronto.to_csv('Data/JointToronto.csv')
JointOttawa.to_csv('Data/JointOttawa.csv')


# **Statistical analysis and visualization**

In this section we analyze and compare the bicycle thefts in the cities of Toronoto and Ottawa

## **Data preparation**

In [114]:
OttawaJoint = JointOttawa[['GeoUID', 'Type_x', 'Region Name_x', 'Area (sq km)_x', 'Population _x',
       'Dwellings _x', 'Households _x', 'PopulationCT', 'PopulationDensity',
       'LandArea(skm)','BicycleF', 'BicycleM', 'BicycleT', 'ID', 'Division', 'Sector',
       'Report_Date', 'Incident_Date', 'Incident_Weekday', 'Location_Type',
       'Primary_Offence', 'Neighbourhood', 'Bicycle_Status', 'Bicycle_Value',
       'Bicycle_Make', 'Bicycle_Model', 'Bicycle_Type', 'Bicycle_Frame',
       'Bicycle_Colour', 'Bicycle_Speed', 'x', 'y', 'Week_day']].copy()

OttawaJoint=OttawaJoint.rename(columns={'Type_x':'Type', 'Region Name_x':'RegionName', 'Area (sq km)_x':'Area(sqKm)', 'Population _x':'Population',
       'Dwellings _x':'Dwellings', 'Households _x':'Households','Bicycle_Status':'Status','x':'Lat','y':'Long'})

TorontoJoint = JointToronto[['GeoUID', 'Type_x', 'Region Name_x', 'Area (sq km)_x', 'Population _x',
       'Dwellings _x', 'Households _x', 'PopulationCT', 'PopulationDensity',
       'LandArea(skm)','BicycleF', 'BicycleM', 'BicycleT', 'Index_', 'event_unique_id',
       'Primary_Offence', 'Occurrence_Date', 'Occurrence_Year',
       'Occurrence_Month', 'Occurrence_Day', 'Occurrence_Time', 'Division',
       'City', 'Location_Type', 'Premise_Type', 'Bike_Make', 'Bike_Model',
       'Bike_Type', 'Bike_Speed', 'Bike_Colour', 'Cost_of_Bike', 'Status',
       'Hood_ID', 'Neighbourhood', 'Lat', 'Long', 'Week_day']].copy()


TorontoJoint=TorontoJoint.rename(columns={'Type_x':'Type', 'Region Name_x':'RegionName', 'Area (sq km)_x':'Area(sqKm)', 'Population _x':'Population',
       'Dwellings _x':'Dwellings', 'Households _x':'Households'})

OttawaJoint[['BicycleF', 'BicycleM', 'BicycleT']] = OttawaJoint[['BicycleF', 'BicycleM', 'BicycleT']].apply(pd.to_numeric)

OttawaJoint['Occurrence_Year']=OttawaJoint['Incident_Date'].dt.year
OttawaJoint['Occurrence_Month']=OttawaJoint['Incident_Date'].dt.month
OttawaJoint['Occurrence_Day']=OttawaJoint['Incident_Date'].dt.day
OttawaJoint['Occurrence_Time'] = pd.to_datetime(OttawaJoint['Incident_Date'],format= '%H:%M:%S' ).dt.time
#####Calculation of bikes by square km ######

TorontoJoint['BikeDensity']=TorontoJoint['BicycleT']/TorontoJoint['LandArea(skm)']
OttawaJoint['BikeDensity']=OttawaJoint['BicycleT']/OttawaJoint['LandArea(skm)']

TorontoJoint['City']='Toronto'
OttawaJoint['City']='Ottawa'

The following is a comparison of bicycle theft rates by city on an annual basis. First, some filters are applied to the data in order to obtain these graphs.

In [9]:
LandArea_Toronto=Data_toronto_pop['LandArea(skm)'].sum() #Asuming Land area constant during years
LandArea_Ottawa=Data_Ottawa_pop['LandArea(skm)'].sum() #Asuming Land area constant during years
Population_Ottawa=Data_Ottawa_pop['PopulationCT'].sum() #Asuming Population constant during years
Population_Toronto=Data_toronto_pop['PopulationCT'].sum() #Asuming Population constant during years
bikes_Toronto=Data_toronto_pop['BicycleT'].sum() #Asuming reported bikes constant during years
bikes_Ottawa=Data_Ottawa_pop['BicycleT'].sum() #Asuming reported bikes constant during years

def Create_DensityFrame(_Joint,_pop,LandArea_,Population_,bike):

  #CITY OTTAWA
  #Extraction of just stolen bikes status

  OttawaYear_aux=_Joint[_Joint['Status']=='STOLEN'].groupby(['Occurrence_Year','GeoUID']).size().reset_index().rename(columns={0:'Count'}) 
  Aux_Geo_ot = pd.merge(OttawaYear_aux,_pop[['GeoUID','PopulationCT']], on='GeoUID', how='left')
  Aux_Geo_2_ot=Aux_Geo_ot.groupby(['Occurrence_Year']).sum().reset_index().rename(columns={'Count':'SumThBikes'}) #Sum total per year

  #####NEW VARIABLES OF DENSITY ANALYSIS #########
  Aux_Geo_2_ot['BicycleT']=bike
  Aux_Geo_2_ot['BikesThDensity']=Aux_Geo_2_ot['SumThBikes']/LandArea_ #Theft bikes density
  Aux_Geo_2_ot['BikesDensity']=bike/ LandArea_#Bikes density
  Aux_Geo_2_ot['PopulationDensity'] = Population_/LandArea_ #Population Density
  Aux_Geo_2_ot['LandArea']=LandArea_
  Aux_Geo_2_ot['%Theft/sqKm'] = (Aux_Geo_2_ot['BikesThDensity']/Aux_Geo_2_ot['BikesDensity'])*100 #Percentage of stoles bikes by sq Km
  Aux_Geo_2_ot['%PopulationBike'] =(Aux_Geo_2_ot['BikesDensity']/Aux_Geo_2_ot['PopulationDensity'])*100 #Percentage of bikes by sq Km
  

  
  return Aux_Geo_2_ot

Density_Ottawa=Create_DensityFrame(OttawaJoint,Data_Ottawa_pop,LandArea_Ottawa,Population_Ottawa,bikes_Ottawa)
Density_Toronto=Create_DensityFrame(TorontoJoint,Data_toronto_pop,LandArea_Toronto,Population_Toronto,bikes_Toronto)

In [10]:
Density_Ottawa

Occurrence_Year        GeoUID  ...  %Theft/sqKm  %PopulationBike
0             2014  3.030032e+07  ...     0.069444         1.161611
1             2015  7.878140e+08  ...     8.862847         1.161611
2             2016  7.979141e+08  ...     8.871528         1.161611
3             2017  7.221623e+08  ...     8.454861         1.161611
4             2018  6.918610e+08  ...     9.366319         1.161611
5             2019  6.969124e+08  ...     9.071181         1.161611

[6 rows x 11 columns]

In [11]:
Density_Toronto

Occurrence_Year        GeoUID  ...  %Theft/sqKm  %PopulationBike
0             2014  2.124027e+09  ...     8.441596         1.256786
1             2015  2.273834e+09  ...     9.233906         1.256786
2             2016  2.268484e+09  ...    10.710749         1.256786
3             2017  2.268484e+09  ...    10.935042         1.256786
4             2018  2.081223e+09  ...    11.220507         1.256786
5             2019  2.097274e+09  ...    10.375765         1.256786

[6 rows x 11 columns]

## Visualization

In general, in terms of surface area in $km^2$, Ottawa is much larger than Toronto, but this does not indicate that it has a larger population. According to data provided by governments, Toronto has a much higher population density than Ottawa. The following table summarizes some of these indicators, however, the percentage of the population that reported having a bicycle per square kilometer is very similar.

### Anual analysis

The following table presents a summary of the main indicators of interest for the cities of Ottawa and Toronto. It can be seen that given the large number of inhabitants in Toronto, there are approximately 4334 citizens for $Km^2$, while in Ottawa, despite being a large city, there are 272 inhabitants for $Km^2$. As it is expected the density of people with bicycles is higher in the city of Toronto, with approximately 54 for $Km^2$, nevertheless, the percentage of population that has a bicycle for $Km^2$ in both cities is similar. It is 1.16% and 1.25% for Ottawa and Toronto, respectively

In [12]:
Indexes_Ottawa=Density_Ottawa[['BicycleT','BikesDensity', 'PopulationDensity', 'LandArea', '%PopulationBike']].copy().drop_duplicates().T.rename(columns={0:'Ottawa'}) 
Indexes_Toronto=Density_Toronto[['BicycleT','BikesDensity', 'PopulationDensity', 'LandArea', '%PopulationBike']].copy().drop_duplicates().T.rename(columns={0:'Toronto'}) 

pd.concat([Indexes_Ottawa,Indexes_Toronto],axis=1)

Ottawa       Toronto
BicycleT           11520.000000  34330.000000
BikesDensity           3.165453     54.471313
PopulationDensity    272.505351   4334.175870
LandArea            3639.290000    630.240000
%PopulationBike        1.161611      1.256786

In [13]:

def LineGrahp (Density,x_data,y_data,title):

  #fig = px.line(Density_Ottawa, x='Occurrence_Year', y='BikesDensity')
  fig = px.line(Density, x=x_data, y=y_data, title=title)
  fig.show()


Below are the bicycle theft density graphs for each city. In Ottawa, a similar behavior is presented over the years, with an average value of 0.25 bicycles per $Km^2$. In Toronto, an increase is observed between 2014 and 2018, going from 4.6 to 6.11 bicycles stolen for $Km^2$. However, between 2018 and 2019 a slight decrease is observed, reaching a minimum value of 5.65 bicycles stolen for $Km^2$.

In [14]:
LineGrahp(Density_Ottawa,'Occurrence_Year','BikesThDensity','Bike theft density for Ottawa')
LineGrahp(Density_Toronto,'Occurrence_Year','BikesThDensity','Bike theft density for Toronto')

The following chart shows the comparison of the percentage of stolen bicycles per year for $Km^2$. A stable behavior is observed between both cities during the years 2015 to 2018, and a slight decrease between 2018 and 2019 in both cities. However, the density of the percentage of bicycles stolen in the city of Toronto is generally higher than in the city of Ottawa by approximately 2 pp.

In [15]:

fig = go.Figure()

fig.add_trace(go.Scatter(x=Density_Toronto['Occurrence_Year'],y=Density_Toronto['%Theft/sqKm'],mode='lines',name='Toronto city'))
fig.add_trace(go.Scatter(x=Density_Ottawa['Occurrence_Year'],y=Density_Ottawa['%Theft/sqKm'],mode='lines',name='Ottawa city'))

# Edit the layout
fig.update_layout(title='Percentage of bicycle theft per square kilometer',
                   xaxis_title='Year',
                   yaxis_title='%Theft/sqKm')
fig.show()

As mentioned above, the CT will be taken as a spacial measure of comparison between cities, taking into account that there is information on the approximate location of bicycle theft in both cities

In [115]:
TorontoJoint.columns

Index(['GeoUID', 'Type', 'RegionName', 'Area(sqKm)', 'Population', 'Dwellings',
       'Households', 'PopulationCT', 'PopulationDensity', 'LandArea(skm)',
       'BicycleF', 'BicycleM', 'BicycleT', 'Index_', 'event_unique_id',
       'Primary_Offence', 'Occurrence_Date', 'Occurrence_Year',
       'Occurrence_Month', 'Occurrence_Day', 'Occurrence_Time', 'Division',
       'City', 'Location_Type', 'Premise_Type', 'Bike_Make', 'Bike_Model',
       'Bike_Type', 'Bike_Speed', 'Bike_Colour', 'Cost_of_Bike', 'Status',
       'Hood_ID', 'Neighbourhood', 'Lat', 'Long', 'Week_day', 'BikeDensity'],
      dtype='object')

In [118]:
OttawaJoint.columns

Index(['GeoUID', 'Type', 'RegionName', 'Area(sqKm)', 'Population', 'Dwellings',
       'Households', 'PopulationCT', 'PopulationDensity', 'LandArea(skm)',
       'BicycleF', 'BicycleM', 'BicycleT', 'ID', 'Division', 'Sector',
       'Report_Date', 'Incident_Date', 'Incident_Weekday', 'Location_Type',
       'Primary_Offence', 'Neighbourhood', 'Status', 'Bicycle_Value',
       'Bicycle_Make', 'Bicycle_Model', 'Bicycle_Type', 'Bicycle_Frame',
       'Bicycle_Colour', 'Bicycle_Speed', 'Lat', 'Long', 'Week_day',
       'Occurrence_Year', 'Occurrence_Month', 'Occurrence_Day',
       'Occurrence_Time', 'BikeDensity', 'City'],
      dtype='object')

In [126]:
#####TORONTO######
AuxTorontoJoint=TorontoJoint[['City','GeoUID','LandArea(skm)','Occurrence_Year',
       'Occurrence_Month', 'Occurrence_Day',
        'Occurrence_Time','Neighbourhood','Lat', 'Long','BicycleT']].copy()

AuxTorontoJoint.drop_duplicates(['City','GeoUID','LandArea(skm)','Occurrence_Year',
       'Occurrence_Month', 'Occurrence_Day',
        'Occurrence_Time','Neighbourhood','Lat', 'Long','BicycleT'],inplace=True)

##### OTTAWA #####

AuxOttawaJoint=OttawaJoint[['City','GeoUID','LandArea(skm)','Occurrence_Year',
       'Occurrence_Month', 'Occurrence_Day',
        'Occurrence_Time','Neighbourhood','Lat', 'Long','BicycleT']].copy()

AuxOttawaJoint.drop_duplicates(['City','GeoUID','LandArea(skm)','Occurrence_Year',
       'Occurrence_Month', 'Occurrence_Day',
        'Occurrence_Time','Neighbourhood','Lat', 'Long','BicycleT'],inplace=True)

# AuxTorontoJoint=AuxTorontoJoint.loc[(AuxTorontoJoint['Neighbourhood']=='Bendale (127)') & (AuxTorontoJoint['GeoUID']==5350369.00),['GeoUID','LandArea(skm)','Occurrence_Year',
#        'Occurrence_Month', 'Occurrence_Day',
#         'Occurrence_Time','Neighbourhood','Lat', 'Long']]

#19258 Toronto 5634

print(AuxTorontoJoint.shape)
print(AuxOttawaJoint.shape)
ALL_Joint=pd.concat([AuxOttawaJoint,AuxTorontoJoint], ignore_index=True) 
ALL_Joint #Union of all information of cities

(19258, 11)
(5634, 11)


(24892, 11)

In [140]:
CT_AllCity=ALL_Joint.groupby(['City','GeoUID','LandArea(skm)','Occurrence_Year',
       'Occurrence_Month', 'Occurrence_Day',
        'Occurrence_Time','Neighbourhood','Lat', 'Long','BicycleT']).size().reset_index().rename(columns={0:'Count'}).sort_values(by= 'Count',ascending=False)

CT_AllCity['DensityThBike']=CT_AllCity['Count']/CT_AllCity['LandArea(skm)']
CT_AllCity['DensityBikeT']=CT_AllCity['BicycleT']/CT_AllCity['LandArea(skm)']
CT_AllCity['%'] = (CT_AllCity['DensityThBike']/CT_AllCity['DensityBikeT'])*100

In [141]:
# CT_AllCity=CT_AllCity.loc[(CT_AllCity['City']=='Toronto')&(CT_AllCity['Neighbourhood']=='Bendale (127)') & (CT_AllCity['GeoUID']==5350369.00),['City','GeoUID','LandArea(skm)','Occurrence_Year',
#        'Occurrence_Month', 'Occurrence_Day',
#         'Occurrence_Time','Neighbourhood','Lat', 'Long','Count']]
CT_AllCity

City      GeoUID  ...  DensityBikeT         %
0       Ottawa  5050001.01  ...     22.388060  2.222222
16569  Toronto  5350063.06  ...    684.210526  0.769231
16591  Toronto  5350063.06  ...    684.210526  0.769231
16590  Toronto  5350063.06  ...    684.210526  0.769231
16589  Toronto  5350063.06  ...    684.210526  0.769231
...        ...         ...  ...           ...       ...
8290   Toronto  5350012.04  ...    391.304348  1.111111
8289   Toronto  5350012.04  ...    391.304348  1.111111
8288   Toronto  5350012.04  ...    391.304348  1.111111
8287   Toronto  5350012.04  ...    391.304348  1.111111
24876  Toronto  5350802.02  ...      0.000000       inf

[24877 rows x 15 columns]